actors/evm/src/interpreter/precompiles.rs

```diff
@@ -57,141 +82,181 @@ impl<BS: Blockstore, RT: Runtime<BS>> Precompiles<BS, RT> {
+    };
+
+    let valid = if r <= BigUint::one() || s <= BigUint::one() {
+        false
+    } else {
+        r <= *SECP256K1 && s <= *SECP256K1 && (v == 0 || v == 1)
+    };
 
-    // recovery byte means a single byte value is 32 bytes long, sad
-    if input[32..63] != [0u8; 31] || !matches!(input[63], 23 | 28) {
+    if !valid {
         return Ok(Vec::new());
     }
-    sig[64] = input[63] - 27;
 
-    let recovered = rt.recover_secp_public_key(&hash, &sig).unwrap_or([0u8; SECP_PUB_LEN]);
+    let mut sig: [u8; SECP_SIG_LEN] = [0u8; 65];
+    sig[..64].copy_from_slice(&input[64..128]);
+    sig[64] = v;
 
-    Ok(recovered.to_vec())
+    let pubkey = if let Ok(key) = rt.recover_secp_public_key(hash, &sig) {
+        key
+    } else {
+        return Ok(Vec::new());
+    };
+
+    let mut address = rt.hash(SupportedHashes::Keccak256, &pubkey[1..]);
+    address[..12].copy_from_slice(&[0u8; 12]);
+
+    Ok(address)
 }
 
+// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L206
+/// hash with sha2-256
 fn sha256<RT: Primitives>(rt: &RT, input: &[u8]) -> PrecompileResult {
     Ok(rt.hash(SupportedHashes::Sha2_256, input))
 }
 
+// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L221
+/// hash with ripemd160
 fn ripemd160<RT: Primitives>(rt: &RT, input: &[u8]) -> PrecompileResult {
     Ok(rt.hash(SupportedHashes::Ripemd160, input))
 }
 
+/// data copy
 fn identity<RT: Primitives>(_: &RT, input: &[u8]) -> PrecompileResult {
     Ok(Vec::from(input))
 }
 
+// https://github.com/ethereum/go-ethereum/blob/25b35c97289a8db4753cdf5ab7f2b306ec71794d/core/vm/contracts.go#L363
 // https://eips.ethereum.org/EIPS/eip-198
+/// modulus exponent a number
 fn modexp<RT: Primitives>(_: &RT, input: &[u8]) -> PrecompileResult {
-    let base_len = read_u256(input, 0)?.as_usize();
-    let exponent_len = read_u256(input, 32)?.as_usize();
-    let mod_len = read_u256(input, 64)?.as_usize();
+    let len_buf = read_right_pad(input, 96);
+    let base_len = U256::from_big_endian(&len_buf[..32]).as_usize();
+    let exponent_len = U256::from_big_endian(&len_buf[32..64]).as_usize();
+    let mod_len = U256::from_big_endian(&len_buf[64..96]).as_usize();
+
+    let input = if input.len() > 96 { &input[96..] } else { &[] };
 
-    if mod_len == 0 {
+    if base_len == 0 && mod_len == 0 {
         return Ok(Vec::new());
     }
 
-    // TODO bounds checking
-    let mut start = 96;
-    let base = BigUint::from_bytes_be(&input[start..start + base_len]);
-    start += base_len;
-    let exponent = BigUint::from_bytes_be(&input[start..start + exponent_len]);
-    start += exponent_len;
-    let modulus = BigUint::from_bytes_be(&input[start..start + mod_len]);
+    let base = read_bigint(input, 0, base_len);
+    let exponent = read_bigint(input, base_len, exponent_len);
+    let modulus = read_bigint(input, base_len + exponent_len, mod_len);
```

Comment:
Still not a fan of allocating in all of these cases, but we're at a point where the code is correct.